syntax = "proto3";

option java_multiple_files = true;
option java_package = "per.elwin.vipuser";
option java_outer_classname = "VipUserProto";
option objc_class_prefix = "VPU";

package vipuser_proto;

service VipUser {
  rpc CreateAcount (CreateAccountRequest) returns (CreateAccountResponse) {}
  rpc Login (LoginRequest) returns (LoginResponse) {}
  rpc Logout (LogoutRequest) returns (LogoutResponse) {}
  rpc ReLogin (ReLoginRequest) returns (ReLoginResponse) {}

  rpc DoBusiness(BusinessRequest) returns (BusinessResponse) {}
}


message AccountInfo {
  string userAlias = 1;
  string passwordSHA256 = 2;
}

message TokenInfo {
  string uuid = 1;
  string accessToken = 2;
  string refreshToken = 3;
}

message Result {
  int32 code = 1;
  string message = 2;
}


message CreateAccountRequest {
  AccountInfo accountInfo = 1;
}

message CreateAccountResponse {
  Result result = 1;
  TokenInfo tokenInfo = 2;
}


message LoginRequest {
  AccountInfo accountInfo = 1;
}

message LoginResponse {
  Result result = 1;
  TokenInfo tokenInfo = 2;
}


message LogoutRequest {
  string accessToken = 1;
}

message LogoutResponse {
  Result result = 1;
}


message ReLoginRequest {
  string refreshToken = 1;
}

message ReLoginResponse {
  Result result = 1;
  TokenInfo tokenInfo = 2;
}



// service BusinessService {
//   rpc DoBusiness(BusinessRequest) returns (BusinessResponse) {}
// }

message UserSession {
  string uuid = 1;;
  string accessToken = 2;
}

message BusinessRequest {
  UserSession session = 1;
  bytes request = 2;
}

message BusinessResponse {
  Result result = 1;
  bytes response = 2;
}
